### DISCUSSION TITLE
A form that lets users apply custom filters to search through a table

### DISCUSSION THREAD

**Discussion by borisfindrik** [sqlpage/SQLPage#536] [Q&A] [answered]
Hi everyone,
how can I substitute empty field in form for % in query?

for table 

```sql
CREATE TABLE envvals (
  date text PRIMARY KEY, 
  temp1 text NOT NULL
);
```

data is like

| date | temp1 |
| --- | --- |
2024-08-09 20:06:35|28.00
2024-08-09 20:11:36|27.00
2024-08-09 20:16:36|28.00
2024-08-09 20:21:36|28.40
2024-08-09 20:31:36|27.80

For form

```sql
select 
    'form'             as component,
    'env_data.sql'     as action;
select 
    'Date and Time' as name,
    'DateTime' as date;
select
    'Temperature' as name, 
    'Temp' as temp1;
```

I try

```sql
select date, temp1 as "temp(C)" 
from envvals
where
      CASE
        WHEN :DateTime IS NOT NULL THEN date like concat ('%',:DateTime,'%') 
        ELSE date like '%' 
      END
and
      CASE
        WHEN :Temp IS NOT NULL THEN temp1 between :Temp and :Temp + 1  
        ELSE temp1 like '%'
      END
order by date;
```

for

```sql
select date, temp1 as "temp(C)"
from envvals
where date like '%2024-08-09%' and temp1 like '%';
```

resulting far more data then expected

```sql
select count(*)
from envvals
where date like '%2024-08-09%' and temp1 like '%';
```

is 288 for 5 measurements per minute

Any input most welcome!

**Comment by lovasoa** [comment, marked as answer]
Hello and welcome to SQLPage !

It looks like you have a problem in your form. You should have `'DateTime' as name` if you want to be able to be able to use `:DateTime` in the next page. You can have a variable name that is different from the displayed label by using the `label` property. You can replace your form queries with

```sql
select 
    'Date and Time' as label,
    'DateTime' as name;
select
    'Temperature' as label, 
    'Temp' as name;
```

### Full working value filtering form example

```sql
select 'form' as component, 
       'Search' as validate, -- Creates a submit button with the text "Search"
       'GET' as method;      -- Form data will be sent as URL parameters

select 
    'Date' as label,         -- Text shown next to the input
    'DateTime' as name,      -- Parameter name used in URL and accessed via $DateTime
    'date' as type,          -- Creates a date picker input
    $DateTime as value;      -- Pre-fills input with existing value from URL

select 'Temperature' as label, 
       'Temp' as name, 
       'number' as type,     -- Creates an input that only accepts numeric values
       $Temp as value;       -- Pre-fills input with existing value from URL

select 'table' as component;

select date, temp1 as "temp(C)"
from envvals
where
      CASE
        WHEN COALESCE($DateTime,'') <> '' THEN  -- Works both when the URL parameter is absent (NULL) and when the field was left blank (empty string coming from the form) 
          date LIKE $DateTime || '%'            -- Matches date prefix, ignoring time
        ELSE true
      END
and
      CASE
        WHEN COALESCE($Temp,'') <> '' THEN     -- Handles NULL and checks for non-empty string
          $Temp <= temp1 AND temp1 < $Temp + 1 -- excludes upper bound
        ELSE true
      END
order by date;
```
Try it live online: https://replit.com/@pimaj62145/SQLPage-custom-value-filtering?v=1

**Reply by borisfindrik** [reply]
Thanks, this look awesome.

**Reply by lovasoa** [reply]
Good! Out of curiosity, what app are you building? Do you have a database connected to temperature sensors ?

**Reply by borisfindrik** [reply]
Naive prediction of rain. I don't think it will work :-) 
I will see results soon after the first major weather change, so far is only hot.
Measuring two temperatures, humidity and pressure with BMP085 and AM2302 connected to esp32 that sends date to mqtt on RPi.
Python script reads the data and write to sqlite db.

**Reply by lovasoa** [reply]
Sounds like a fun project! Please share your write-up on this forum if you do one when you have your results!

**Comment by saitou-yuki** [comment]
- 前営業日にやったこと
  - 定常タスク
    - PRレビュー
  - ネイティブアプリ
    - 商品詳細 - キャンペーン・クーポン販促枠API定義修正
      - https://daytonajp.backlog.com/view/APP_RENEWAL-3286#comment-636018510
    - https://github.com/daytona-international/di-ddp/issues/27014
      - https://github.com/daytona-international/di-ddp/issues/27015
    - https://github.com/daytona-international/di-ddp/issues/27215
      - https://github.com/daytona-international/di-ddp/issues/27216
- 今日やること
  - ネイティブアプリ
    - https://github.com/daytona-international/di-ddp/issues/27014
      - [ ] https://github.com/daytona-international/di-ddp/issues/27018
      - [ ] https://github.com/daytona-international/di-ddp/issues/27219
    - https://github.com/daytona-international/di-ddp/issues/27215
      - [ ] https://github.com/daytona-international/di-ddp/issues/27216
      - [ ] https://github.com/daytona-international/di-ddp/issues/27217
      - [ ] https://github.com/daytona-international/di-ddp/issues/27218
      - [ ] https://github.com/daytona-international/di-ddp/issues/27317
    - 残タスク
      - [ ] https://github.com/daytona-international/di-ddp/issues/27220
    - 保留中
      - [ ] ZETA_SEARCH:価格帯検索時の参照項目について確認
        - [To: ZETA様](https://daytonajp.backlog.com/view/APP_RENEWAL-2842)
- 困っていること
- 滞留PR報告
  - di-docs
    - https://github.com/daytona-international/di-docs/pull/1989
